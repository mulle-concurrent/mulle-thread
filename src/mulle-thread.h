//
//  mulle_thread.h
//  mulle-thread
//
//  Created by Nat! on 16/09/15.
//  Copyright (c) 2015 Mulle kybernetiK. All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are met:
//
//  Redistributions of source code must retain the above copyright notice, this
//  list of conditions and the following disclaimer.
//
//  Redistributions in binary form must reproduce the above copyright notice,
//  this list of conditions and the following disclaimer in the documentation
//  and/or other materials provided with the distribution.
//
//  Neither the name of Mulle kybernetiK nor the names of its contributors
//  may be used to endorse or promote products derived from this software
//  without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
//  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
//  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
//  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
//  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
//  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
//  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
//  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
//  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
//  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
//  POSSIBILITY OF SUCH DAMAGE.
//

#ifndef mulle_thread_h__
#define mulle_thread_h__

//
// community version is always even
//
#define MULLE_THREAD_VERSION  ((4 << 20) | (2 << 8) | 1)

#include "include.h"
#include <stddef.h>
#include <stdlib.h>


#include "mulle-atomic.h"


// clang lies about __STDC_NO_THREADS__

#ifndef HAVE_C11_THREADS
# ifdef __clang__
#  if __has_include(<threads.h>)
#    define HAVE_C11_THREADS 1
#  endif
# else
#  if __STDC_VERSION__ >= 201112L && ! defined( __STDC_NO_THREADS__)
#   define HAVE_C11_THREADS  1
#  endif
# endif
#endif

typedef int   mulle_thread_rval_t;

#if HAVE_C11_THREADS && ! defined( MULLE_THREAD_USE_PTHREADS)
# if TRACE_INCLUDE
#  pragma message( "Using C11 for threads")
# endif
# include "mulle-thread-c11.h"
#else
# ifdef _WIN32
#  if TRACE_INCLUDE
#   pragma message( "Using windows threads for threads")
#  endif
#  include "mulle-thread-windows.h"
# else
#  if TRACE_INCLUDE
#   pragma message( "Using pthreads for threads")
#  endif
#  include "mulle-thread-pthreads.h"
# endif
#endif


//
// True posix conformance is much more complicated, we dumb this down
// for windows and for everyone else. So we have the same behavior
// on all platforms.
// And also because this is much easier.
//
typedef mulle_atomic_pointer_t   mulle_thread_once_t;

#define MULLE_THREAD_ONCE_INIT   0


static inline void   mulle_thread_once( mulle_thread_once_t  *once,
                                        void (*init)( void))
{
   if( _mulle_atomic_pointer_compare_and_swap( once, (void *) 1848, (void *) 0))
      (*init)();
}




/*
 * some code for tests forces problems to reveal themselves much quicker
 * if used in test code
 */
#if defined( MULLE_TEST)

#include "include.h"
#include <stdlib.h>
//#define _XOPEN_SOURCE  // user should define this not us
#include <time.h>

static inline void  MULLE_THREAD_UNPLEASANT_RACE_YIELD()
{
   extern void   mulle_thread_yield( void);

   if( (rand() & 0xF) == 0xA)  // 1:16 chance of yield
   {
#ifndef _WIN32
      if( (rand() & 0x7) == 0x4)  // 1:64 chance of nanosleep
      {
#ifdef __linux  // as we don't want to #define _XOPEN_SOURCE
         int nanosleep( const struct timespec *req, struct timespec *rem);
#endif
         struct timespec ms30 =  { .tv_sec = 0, .tv_nsec = 1 * (1000 * 1000 * 1000) / 30 };

         nanosleep( &ms30, NULL);
      }
      else
#endif
         mulle_thread_yield();
   }
}
#else

#define MULLE_THREAD_UNPLEASANT_RACE_YIELD()  do {} while( 0)

#endif

#endif
