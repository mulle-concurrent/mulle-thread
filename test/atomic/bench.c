//
//  main.c
//  test-atomic
//
//  Created by Nat! on 19.03.15.
//  Copyright (c) 2015 Mulle kybernetiK. All rights reserved.
//
//  Redistributions of source code must retain the above copyright notice, this
//  list of conditions and the following disclaimer.
//
//  Redistributions in binary form must reproduce the above copyright notice,
//  this list of conditions and the following disclaimer in the documentation
//  and/or other materials provided with the distribution.
//
//  Neither the name of Mulle kybernetiK nor the names of its contributors
//  may be used to endorse or promote products derived from this software
//  without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
//  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
//  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
//  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
//  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
//  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
//  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
//  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
//  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
//  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
//  POSSIBILITY OF SUCH DAMAGE.
//
#include <mulle-thread/mulle-thread.h>
#include <assert.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


#define PROGRESS        0
#define MAX_ITERATIONS  100
#define LOOPS           10000
#define MAX_THREADS     4


typedef double mulle_timeinterval_t;
typedef double mulle_relativetime_t;

#ifdef _WIN32

#include <windows.h>

// https://stackoverflow.com/questions/5404277/porting-clock-gettime-to-windows

mulle_timeinterval_t   mulle_timeinterval_now( void)
{
   __int64   wintime;

   GetSystemTimeAsFileTime((FILETIME*)&wintime);
   wintime -=116444736000000000i64;  //1jan1601 to 1jan1970
   return( (mulle_timeinterval_t) wintime / 10000000.0);
}

#else

mulle_timeinterval_t   mulle_timeinterval_now( void)
{
   struct timespec   now;

   clock_gettime( CLOCK_REALTIME, &now);
   return( now.tv_sec + now.tv_nsec / (1000.0*1000*1000));
}

#endif



static mulle_thread_tss_t   name_thread_key;
char  *mulle_aba_thread_name( void);


static mulle_atomic_pointer_t    central;

#pragma mark - run test

static void    run_atomic_read_cas( void)
{
   unsigned long   i;
   void            *value;
   void            *expect;

   for( i = 0; i < LOOPS; i++)
   {
      do
      {
         expect = _mulle_atomic_pointer_read( &central);
         value  = (void *) ((intptr_t) expect + 1);
      }
      while( ! _mulle_atomic_pointer_weakcas( &central, value, expect));
   }
}


static void    run_atomic_better_read_cas( void)
{
   unsigned long   i;
   void            *value;
   void            *expect;
   void            *actual;

   for( i = 0; i < LOOPS; i++)
   {
      expect = _mulle_atomic_pointer_read( &central);
      for(;;)
      {
         value  = (void *) ((intptr_t) expect + 1);
         actual = __mulle_atomic_pointer_cas( &central, value, expect);
         if( actual == expect)
            break;
         expect = actual;
      }
   }
}


static void    run_atomic_double_cas( void)
{
   unsigned long   i;
   void            *value;
   void            *expect;
   void            *prev;

   expect = _mulle_atomic_pointer_read( &central);
   for( i = 0; i < LOOPS; i++)
   {
      for(;;)
      {
         value = (void *) ((intptr_t) expect + 1);
         prev  = __mulle_atomic_pointer_weakcas( &central, value, expect);
         if( prev == expect)
            break;
         expect = prev;
      }
   }
}


static void   _wait_around( mulle_atomic_pointer_t *n_threads)
{
   _mulle_atomic_pointer_decrement( n_threads);
   while( _mulle_atomic_pointer_read( n_threads) != 0)
      mulle_thread_yield();
}


struct thread_info
{
   char                       name[ 64];
   void                       (*f)( void);
   mulle_atomic_pointer_t     *n_threads;
};


static mulle_thread_rval_t   run_test( struct thread_info *info)
{
   mulle_thread_tss_set( name_thread_key, strdup( info->name));

   // all thread should start at same time...
   _wait_around( info->n_threads);
   (*info->f)();

   return( 0);
}


static void   finish_test( void)
{
}


void  multi_threaded_test( intptr_t n, void (*f)( void))
{
   int                      i;
   mulle_thread_t           *threads;
   struct thread_info       *info;
   mulle_atomic_pointer_t   n_threads;

   threads = alloca( sizeof( mulle_thread_t) * n);
   {
      assert( threads);

      _mulle_atomic_pointer_nonatomic_write( &n_threads, (void *) n);

      info = alloca( sizeof(struct thread_info) * n);
      {
         _mulle_atomic_pointer_nonatomic_write( &central, (void *) 0);

         for( i = 1; i < n; i++)
         {
            info[ i].n_threads = &n_threads;
            info[ i].f         = f;
            sprintf( info[ i].name, "thread #%d", i);

            if( mulle_thread_create( (void *) run_test, &info[ i], &threads[ i]))
               abort();
         }

         info[ 0].n_threads = &n_threads;
         info[ 0].f         = f;
         sprintf( info[ 0].name, "thread #%d", 0);
         run_test( &info[ 0]);

         for( i = 1; i < n; i++)
            if( mulle_thread_join( threads[ i]))
            {
               perror( "mulle_thread_join");
               abort();
            }

         finish_test();
      }
   }
}


char  *mulle_aba_thread_name( void)
{
   return( mulle_thread_tss_get( name_thread_key));
}



mulle_relativetime_t   test( void (*f)( void))
{
   unsigned long          i;
   unsigned int           j;
   int                    rval;
   mulle_timeinterval_t   start;
   mulle_timeinterval_t   end;

   start = mulle_timeinterval_now();

   for( i = 1; i <= MAX_ITERATIONS; i++)
   {
      for( j = 1; j <= MAX_THREADS; j += j)
      {
         multi_threaded_test( j, f);
      }
   }

   end = mulle_timeinterval_now();

   return( end - start);
}


int   main( int argc, const char * argv[])
{
   int                    rval;
   mulle_relativetime_t   elapsed_read_cas;
   mulle_relativetime_t   elapsed_double_cas;
   mulle_relativetime_t   elapsed_better_cas;

   rval = mulle_thread_tss_create( free, &name_thread_key);
   assert( ! rval);

   rval = mulle_thread_tss_set( name_thread_key, strdup( "main"));
   assert( ! rval);

   test( run_atomic_better_read_cas);
   test( run_atomic_read_cas);
   test( run_atomic_double_cas);

   elapsed_better_cas = test( run_atomic_better_read_cas);
   fprintf( stderr, "better-cas %g\n", elapsed_better_cas);

   elapsed_read_cas   = test( run_atomic_read_cas);
   fprintf( stderr, "read-cas %g\n", elapsed_read_cas);

   elapsed_double_cas = test( run_atomic_double_cas);
   fprintf( stderr, "double-cas %g\n", elapsed_double_cas);

   printf( "better-cas is %s than read-cas, which is %s than double-cas\n",
            elapsed_better_cas < elapsed_read_cas ? "faster" : "slower",
            elapsed_read_cas < elapsed_double_cas ? "faster" : "slower");


   return( 0);
}
