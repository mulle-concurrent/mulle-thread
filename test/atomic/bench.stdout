better-cas is faster than read-cas, which is faster than double-cas
